### PR TITLE
Folder permissions for CKEditor missing allow access permissions

### LINKED ISSUE #4573: Resource Manager permissions bug/issue in DNN 9.8.1 and above (HTML Editor Browse Server showing all folders)
I believe the Resource Manager has a folder permissions bug/issue. The HTML Editor "Browse Server" is showing ALL folders to Roles after upgrading from DNN 9.7.2 to 9.8.1 no matter how the permissions are set in the Resource Manager.

When using the HTML Editor, selecting the image or file icon, selecting the "Browse Server" button. The Folder Tree shows all folders no matter what the folder permissions are set to for Roles in the Resource Manager. In DNN 9.7.2 I had folder permissions set so that I could hide some of the folders from certain Roles, but after upgrading to 9.8.1 it shows all the folders.

I have tried different folder permission settings in the PB > Site Assets, but nothing works. I have even tried replacing with the new Resource Manger but that did not work either.

I have tested this on a test production site and a test dnndev.me site and both have the same issue.

**To replicate the issue. Go to the Site Assets and assign folder permissions on a folder to hide it from a "Role". Then login as a user assigned to that Role and use the HTML module "Browse Server" and you will see the folder you are trying to hide.**

I have attached a couple of screen shots.

DNN 9 8 1 Site Assets Folder permissions

DNN 9 8 1 HTML Editor Browse Server showing all folders

Thanks,
Craig

### PR DESCRIPTION
**Pull request by yog-it**
Fixes #4573

## Summary
`GetFoldersByPermissions` requires access permissions to the specific permissions per UserID or Role same as deny permissions.

### CODE REVIEW

**Review by sleupold — COMMENTED** [review]
You should not rely on common role names - those might get translated and be different than "Administrators".

**Review by sleupold — COMMENTED** [review]
- You should not rely on common role names - those might get translated and be different than "Administrators".
- You should name your temp table appropriately, "#Granted_Folders", not "#Skip_Folders" 
- there is no permission check for Host folders included (should be for superusers only)
- for performance reasons, I'd avoid using a temp table
- some checks seem to be duplicate

**Comment by yog-it** [comment]
@sleupold thank you.  The SP was previously written that way, but you are correct, it should not be there.

**Comment by yog-it** [comment]
I made the change for the AdministratorRoleId.  I'll look over the rest of your points to correct it.  It does appear to be an SP that needed more updating, then just fixing the permissions issues in #4573.

**Review by sleupold — COMMENTED** [review]
Just from reviewing the SQL code, I am not sure, deny permission are handled properly.

**Comment by sleupold** [comment]
I just compared with the version from my TurboScript, and noted an glitch in there aa well - caused by the compexity, how permissions are stored and inappropriate construction of the permission param of this sproc (not your fault, of course).

**Comment by sleupold** [comment]
Just for comparison, my current solution (though not perfect):

CREATE PROCEDURE {databaseOwner}[{objectQualifier}GetFoldersByPermissions]
    @PortalID    Int,            -- Null|-1 retrieves host folders
    @Permissions nVarChar(300),  -- list of permissions, to be required any of it, not the complete set. Use '*', '' OR Null for ANY.
    @UserID      Int,            -- Id of the user to be inspected, Null|-1 for anonymous users
    @FolderId    Int,            -- optional ID   of a single folder to check (or Null|-1 for all)
    @FolderPath  nVarChar(300)   -- optional path of a single folder to check (or Null|'' for all)
AS
BEGIN
    DECLARE @IsSuperUser  Bit = 0
    DECLARE @IsAdmin      Bit = 0
    DECLARE @AllPerm      Int = 0
    DECLARE @ReadPerm     Int = 0
    DECLARE @WritePerm    Int = 0
    DECLARE @BrowsePerm   Int = 0
    DECLARE @AddPerm      Int = 0

    -- Determine Admin or SuperUser
    IF IsNull(@UserID, -1) > 0 BEGIN -- not anonymous user
        SELECT @IsSuperUser = IsSuperUser FROM {databaseOwner}[{objectQualifier}Users] WHERE UserId = @UserID;
        IF @IsSuperUser != 0
            SET @IsAdmin = 1; -- superuser has portal admin permission as well 
         ELSE IF IsNull(@PortalID, -1) = -1
            SET @IsAdmin = 0 -- only superusers may access host files
         ELSE IF {databaseOwner}[{objectQualifier}UserIsInRole](@UserID, {databaseOwner}[{objectQualifier}AdministratorRoleId](@PortalID)) != 0
            SET @IsAdmin = 1;
    END;

    IF (IsNull(@PortalID, -1) = -1 AND @IsSuperUser != 0) OR (@IsAdmin != 0)
        -- shortcut for admins/superusers: return all folders (they are having always access to all of them)
        SELECT F.*
         FROM  {databaseOwner}[{objectQualifier}Folders] F
         WHERE IsNull(F.PortalID, -1) = IsNull(@PortalID, -1)
           AND (F.FolderID   = @FolderID   OR IsNull(@FolderID,    -1) =  -1)
           AND (F.FolderPath = @FolderPath OR IsNull(@FolderPath, N'') = N'')
         ORDER BY F.PortalID, F.FolderPath
         OPTION (OPTIMIZE FOR (@PortalID UNKNOWN));
    ELSE BEGIN
        --Retrieve Permission IDs
            SELECT TOP 1 @AllPerm    = PermissionID FROM {databaseOwner}[{objectQualifier}Permission] WHERE PermissionCode = N'SYSTEM_FOLDER' AND ModuleDefId = -1 AND PermissionKey = N'WRITE';
        IF IsNull(@Permissions, N'') LIKE N'%WRITE%'  SET @WritePerm = @AllPerm -- always checked, because WRITE always rules all other permissions
        IF IsNull(@Permissions, N'') LIKE N'%READ%'
            SELECT TOP 1 @ReadPerm   = PermissionID FROM {databaseOwner}[{objectQualifier}Permission] WHERE PermissionCode = N'SYSTEM_FOLDER' AND ModuleDefId = -1 AND PermissionKey = N'READ';
        IF IsNull(@Permissions, N'') LIKE N'%BROWSE%'
            SELECT TOP 1 @BrowsePerm = PermissionID FROM {databaseOwner}[{objectQualifier}Permission] WHERE PermissionCode = N'SYSTEM_FOLDER' AND ModuleDefId = -1 AND PermissionKey = N'BROWSE';
        IF IsNull(@Permissions, N'') LIKE N'%ADD%' -- advanced permission provider only
            SELECT TOP 1 @AddPerm    = PermissionID FROM {databaseOwner}[{objectQualifier}Permission] WHERE PermissionCode = N'SYSTEM_FOLDER' AND ModuleDefId = -1 AND PermissionKey = N'ADD';

        SELECT * FROM {databaseOwner}[{objectQualifier}Folders]
         WHERE (PortalID = IsNull(@PortalID, -1))
           AND (FolderPath = @FolderPath OR IsNull(@FolderPath, N'') = N'')
           AND (FolderID   = @FolderID   OR IsNull(@FolderID, -1) = -1)
           AND FolderId IN -- granted folders
                (SELECT DISTINCT P.FolderId FROM {databaseOwner}[{objectQualifier}FolderPermission] P
                                       LEFT JOIN {databaseOwner}[{objectQualifier}UserPortalRoles](@PortalID, @UserID) R ON P.RoleID = R.RoleID
                  WHERE (P.UserID = @UserID OR (R.RoleID Is Not Null))
                    AND PermissionID IN (@AddPerm, @ReadPerm, @BrowsePerm, @AllPerm)
                    AND AllowAccess = 1)
           AND NOT FolderId IN -- denied folders
                (SELECT DISTINCT P.FolderId FROM {databaseOwner}[{objectQualifier}FolderPermission] P
                                       LEFT JOIN {databaseOwner}[{objectQualifier}UserPortalRoles](@PortalID, @UserID) R ON P.RoleID = R.RoleID
                  WHERE (P.UserID = @UserID OR (R.RoleID Is Not Null))
                    AND PermissionID IN (@AddPerm, @ReadPerm, @BrowsePerm, @WritePerm)
                    AND AllowAccess = 0)
         ORDER BY PortalID, FolderPath
         OPTION (OPTIMIZE FOR (@FolderPath Unknown, @FolderId Unknown, @UserID Unknown));
    END
END -- Procedure
GO

**Comment by yog-it** [comment]
Yes, this is much more efficient and nicer looking 😄.  I could not find the function `UserPortalRoles`, I'm I right on this implementation: 
```
-- User roles per portal function
CREATE FUNCTION {databaseOwner}[{objectQualifier}UserPortalRoles]
(
	@PortalId Int,
	@UserId Int
)
RETURNS @returnUserRoles TABLE 
(
	RoleID int PRIMARY KEY NOT NULL
)
AS 
BEGIN
	INSERT INTO @returnUserRoles (RoleID) VALUES (-1) --Add -1 for "All Users" 
	INSERT @returnUserRoles
	SELECT UR.RoleID FROM {databaseOwner}[{objectQualifier}UserRoles] UR
	JOIN {databaseOwner}[{objectQualifier}Roles] R ON UR.RoleID = R.RoleID AND PortalID = @PortalId
	WHERE UR.UserID = @UserId
RETURN;
END;
GO
```

**Comment by sleupold** [comment]
sorry, it's part of my [Turbo Scripts](https://github.com/dnnwerk/dnnScript)
'
-- =============================================
-- Author:		Sebastian Leupold, dnnWerk
-- Create date: 2017-10-17
-- Description:	Return all Roles of a User
--              for Permission Checks
-- Note:        IsSuperuser not evaluated (perf)
-- =============================================
CREATE FUNCTION {databaseOwner}[{objectQualifier}UserPortalRoles]
(
   @PortalID int, -- not Null 
   @UserID   int  -- not Null, -1 for anonymous users | > 0 for real users
)
RETURNS TABLE 
AS
RETURN 
(  SELECT RoleID 
    FROM  {databaseOwner}[{objectQualifier}vw_UserRoles]
    WHERE PortalID = @PortalID
      AND UserID   = @UserID
       AND ISNull(EffectiveDate, GetDate()) <= GetDate()
       AND IsNull(ExpiryDate,    GetDate()) >= GetDate()
   UNION (SELECT -1)
   UNION (SELECT -3 WHERE @UserID <= 0)
)
GO
'
based on View
'
CREATE VIEW {databaseOwner}[{objectQualifier}vw_UserRoles]
AS
    SELECT UR.UserRoleID,
           R.PortalID,
           R.RoleID,
           R.RoleName,
           R.Description,
           U.UserID,
           U.Username,
           U.DisplayName,
           U.Email,
           UR.Status,
           UR.IsOwner,
           R.SecurityMode,
           R.ServiceFee,
           R.BillingFrequency,
           R.TrialPeriod,
           R.TrialFrequency,
           R.BillingPeriod,
           R.TrialFee,
           R.IsPublic,
           R.AutoAssignment,
           R.RoleGroupID,
           R.RSVPCode,
           R.IconFile,
           R.IsSystemRole,
           UR.EffectiveDate,
           UR.ExpiryDate,
           UR.IsTrialUsed,
           UR.CreatedByUserID,
           UR.CreatedOnDate,
           UR.LastModifiedByUserID,
           UR.LastModifiedOnDate
     FROM  {databaseOwner}[{objectQualifier}UserRoles] AS UR
     JOIN  {databaseOwner}[{objectQualifier}Users]     AS U ON UR.UserID = U.UserID
     JOIN  {databaseOwner}[{objectQualifier}Roles]     AS R ON UR.RoleID = R.RoleID
     WHERE R.Status = 1
GO'

**Comment by valadas** [comment]
It would be nice to have manageable, easier to review, small PRs to bring nice stuff from those turboscripts into the platform.

**Comment by mitchelsellers** [comment]
@sleupold Can you confirm that you are ok with this being merged as is?  Since it has been updated to use the content from your set of scripts.  We will call attention to your contribution as part of this, even as you are not the actual submitter of this PR.

File: DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider
Comment (mitchelsellers) [inline DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider:84]: This needs to be wrapped in a conditional

**Review by mitchelsellers — CHANGES_REQUESTED** [review]

File: DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider
Comment (yog-it) [reply DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider]: Thank you @mitchelsellers What is the minimum version of SQL the TSQL scripts for DNN need to be?  The reason I ask is, there is an [[IF EXISTS] clause to the DROP FUNCTION ](https://docs.microsoft.com/en-us/sql/t-sql/statements/drop-function-transact-sql?view=sql-server-ver15) with the statement "Available beginning with SQL Server 2016 and in SQL Database" in the Argument section.

Comment (bdukes) [reply DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider]: We [currently still support](https://www.dnndocs.com/content/getting-started/setup/requirements/index.html#database-server) SQL Server 2008 R2.

We may want to look at adjusting that at some point, though. [Entity Framework Core supports SQL Server 2012+](https://docs.microsoft.com/en-us/ef/core/providers/sql-server/#supported-database-engines). It looks like [Entity Framework 6 supports older SQL Server](https://stackoverflow.com/a/25240352/2688) (back to at least 2000).

Comment (mitchelsellers) [reply DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider]: I would say for sure we keep it to our current minimums

**Comment by mitchelsellers** [comment]
@yog-it One last request if you can, can you remove the comments from before the statements in the SQL?  It looks like I cannot push changes to your branch.

File: DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider
Comment (sleupold) [reply DNN Platform/Website/Providers/DataProviders/SqlDataProvider/09.09.02.SqlDataProvider]: Extended Support for SQL Server 2008 R2 latet SU will end Juli 2022, and usually we drop support after that.
However, users are advised to use a product version with mainstream support - and given the fact that SQL Server Express Edition is sufficient for most small to medium DNN websites, there is in fact no reason not to upgrade or replace with a current release of SQL Server.

**Review by mitchelsellers — APPROVED** [review]

**Review by bdukes — APPROVED** [review]

**Comment by bdukes** [comment]
@sleupold we're ready to merge this once we get the okay from you, since you authored the main content. Thanks!

**Comment by bdukes** [comment]
@sleupold 👆🏻

**Comment by mitchelsellers** [comment]
@sleupold can we get your approval on this inclusion of this?

**Comment by mitchelsellers** [comment]
At this point in time, after discussion with @sleupold at the DNN TAG meeting, it was determined that he was not comfortable with this PR progressing, in its current form due to issues/limitations in how this was implemented.  It can be re-submitted in the future once those have been addressed, and approval from the initial source of the contribution.

**Comment by yog-it** [comment]
Understandable.  Thanks @mitchelsellers.

**Comment by cgentryls** [comment]
So where does this issue stand, will it be resolved in DNN 9.10.0?
I cannot upgrade our website from DNN 9.7.2 because of this issue which is causing my original issue #4573

**Comment by valadas** [comment]
@sleupold I don't remember the details of the discussion, you have any input on why this was not correct and what should be correct ?

**Comment by cgentryls** [comment]
@valadas you had added this to the 9.10.0 milestone on Jul 7, but from the 9.10.0 release notes, it does not look like it was fixed. Are there any plans to fix this issue? 
I cannot upgrade our website from DNN 9.7.2 because of this issue which is causing my original issue #4573

**Comment by valadas** [comment]
After that we changed how we assign milestones to issues, we now use major/minor/patch to mark in which kind of release this could go if we do receive a pull request contribution. We cannot guarantee any specific version as it all depends if a contribution comes in, so this way it makes it a bit less maintenance than trying to predict in which version it will get in.

**Comment by valadas** [comment]
Oh, this is the pull request and not the issue. This pull request was refused as per the discussion above. Issue #4573 is still open.

**Pull request closed without merging**
